airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StatePersistence.java

```diff
@@ -0,0 +1,323 @@
+    if (state != null) {
+      boolean hasState = ctx.selectFrom(STATE)
+          .where(
+              STATE.CONNECTION_ID.eq(connectionId),
+              isNullOrEquals(STATE.STREAM_NAME, streamName),
+              isNullOrEquals(STATE.NAMESPACE, namespace))
+          .fetch().isNotEmpty();
+
+      final JSONB jsonbState = JSONB.valueOf(Jsons.serialize(state));
+      final OffsetDateTime now = OffsetDateTime.now();
+
+      if (!hasState) {
+        ctx.insertInto(STATE)
+            .columns(
+                STATE.ID,
+                STATE.CREATED_AT,
+                STATE.UPDATED_AT,
+                STATE.CONNECTION_ID,
+                STATE.STREAM_NAME,
+                STATE.NAMESPACE,
+                STATE.STATE_,
+                STATE.TYPE)
+            .values(
+                UUID.randomUUID(),
+                now,
+                now,
+                connectionId,
+                streamName,
+                namespace,
+                jsonbState,
+                convertStateType(stateType))
+            .execute();
+
+      } else {
+        ctx.update(STATE)
+            .set(STATE.UPDATED_AT, now)
+            .set(STATE.STATE_, jsonbState)
+            .where(
+                STATE.CONNECTION_ID.eq(connectionId),
+                isNullOrEquals(STATE.STREAM_NAME, streamName),
+                isNullOrEquals(STATE.NAMESPACE, namespace))
+            .execute();
+      }
+
+    } else {
+      // If the state is null, we remove the state instead of keeping a null row
+      ctx.deleteFrom(STATE)
+          .where(
+              STATE.CONNECTION_ID.eq(connectionId),
+              isNullOrEquals(STATE.STREAM_NAME, streamName),
+              isNullOrEquals(STATE.NAMESPACE, namespace))
+          .execute();
+    }
+  }
+
+  /**
+   * Helper function to handle null or equal case for the optional strings
+   *
+   * We need to have an explicit check for null values because NULL != "str" is NULL, not a boolean.
+   *
+   * @param field the targeted field
+   * @param value the value to check
+   * @return The Condition that performs the desired check
+   */
+  private static Condition isNullOrEquals(final Field<String> field, final String value) {
+    return value != null ? field.eq(value) : field.isNull();
+  }
+
+  /**
+   * Get the StateType for a given list of StateRecords
+   *
+   * @param connectionId The connectionId of the records, used to add more debugging context if an
+   *        error is detected
+   * @param records The list of StateRecords to process, must not be empty
+   * @return the StateType of the records
+   * @throws IOException If StateRecords have inconsistent types
+   */
+  private io.airbyte.db.instance.configs.jooq.generated.enums.StateType getStateType(final UUID connectionId, final List<StateRecord> records) {
+    final List<io.airbyte.db.instance.configs.jooq.generated.enums.StateType> types = records.stream().map(r -> r.type).distinct().toList();
```

Comment:
would it make sense to just use a set?

Reply:
I lean towards the set, because when the variable gets used later down I look at the type and by seeing set I know it'll be unique values, while it is easy to miss the `distinct` when scanning the code. also agree, this is moving into style though.